server/api.js

```diff
@@ -0,0 +1,551 @@
+    } else {
+      res.status(400).send({ status: 'error', message: 'Bad Request' })
+      helpers.logError(`Bad request to ${req.path}: ${validationErrors.array()}`)
+    }
+  } catch (error) {
+    res.status(500).send({ status: 'error', message: 'Internal Server Error' })
+    helpers.logError(`Internal server error at ${req.path}: ${error.message}`)
+  }
+}
+
+const validateUpdateClient = [
+  Validator.header(['Authorization']).notEmpty(),
+  Validator.param(['clientId']).notEmpty(),
+  Validator.body(['displayName', 'fromPhoneNumber', 'language']).trim().isString().notEmpty(),
+  Validator.body(['responderPhoneNumbers', 'fallbackPhoneNumbers', 'heartbeatPhoneNumbers', 'incidentCategories']).isArray(),
+  Validator.body(['reminderTimeout', 'fallbackTimeout']).isInt({ min: 0 }),
+  Validator.body(['isDisplayed', 'isSendingAlerts', 'isSendingVitals']).isBoolean(),
+]
+
+async function handleUpdateClient(req, res) {
+  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
+
+  try {
+    if (validationErrors.isEmpty()) {
+      const client = await db.getClientWithId(req.params.clientId)
+
+      // check that the client exists
+      if (client == null) {
+        res.status(404).send({ status: 'error', message: 'Not Found' })
+
+        return
+      }
+
+      // attempt to update the client
+      const updatedClient = await db.updateClient(
+        req.body.displayName,
+        req.body.fromPhoneNumber,
+        req.body.responderPhoneNumbers,
+        req.body.reminderTimeout,
+        req.body.fallbackPhoneNumbers,
+        req.body.fallbackTimeout,
+        req.body.heartbeatPhoneNumbers,
+        req.body.incidentCategories,
+        req.body.isDisplayed,
+        req.body.isSendingAlerts,
+        req.body.isSendingVitals,
+        req.body.language,
+        req.params.clientId,
+      )
+
+      // something bad happened and the client wasn't updated; blame it on the request
+      if (updatedClient == null) {
+        res.status(400).send({ status: 'error', message: 'Bad Request' })
+
+        return
+      }
+
+      res.status(200).send({ status: 'success', data: updatedClient })
+    } else {
+      res.status(400).send({ status: 'error', message: 'Bad Request' })
+      helpers.logError(`Bad request to ${req.path}: ${validationErrors.array()}`)
+    }
+  } catch (error) {
+    res.status(500).send({ status: 'error', message: 'Internal Server Error' })
+    helpers.logError(`Internal server error at ${req.path}: ${error.message}`)
+  }
+}
+
+const validateUpdateClientButton = [
+  Validator.header(['Authorization']).notEmpty(),
+  Validator.param(['clientId', 'buttonId']).notEmpty(),
+  Validator.body(['displayName', 'phoneNumber', 'buttonSerialNumber']).trim().isString().notEmpty(),
+  Validator.body(['isDisplayed', 'isSendingAlerts', 'isSendingVitals']).isBoolean(),
+]
+
+async function handleUpdateClientButton(req, res) {
```

Comment:
Fun use case for you... what if we want to move a button from one client to another. This could happen in some of our clients who have multiple locations with different responder phone numbers who decide they want to rearrange their buttons for some reason.

The current Sensors dashboard lets users do this using the Client dropdown menu:
![image](https://github.com/bravetechnologycoop/BraveButtons/assets/1490437/eae5bc64-f10d-4018-a078-38f9dc0329f2)

What do you think about adding a `clientId` parameter to this and the db.updateButton call?

Ditto for gateways.

Reply:
Done for both of these! I agree with having the option to move buttons/gateways from one client to another.